tools/skbstat.py

```diff
@@ -0,0 +1,1079 @@
+}
+
+static inline int sk_ip_match(struct sock *sk, struct filter *filter,
+			      union filter_args *args)
+{
+	u32 saddr, daddr;
+	u16 sport, dport;
+
+	if (args->input) {
+		// see kernel source code include/net/inet_hashtables.h for detail.
+		saddr = sk->sk_daddr;
+		daddr = sk->sk_rcv_saddr;
+		sport = sk->sk_dport;
+		dport = sk->sk_num;
+	} else if (args->output) {
+		daddr = sk->sk_daddr;
+		saddr = sk->sk_rcv_saddr;
+		dport = sk->sk_dport;
+		sport = sk->sk_num;
+	} else {
+		goto miss;
+	}
+
+	if (ip_addr_match(saddr, daddr, filter)) {
+		return port_match(htons(sport), htons(dport), filter);
+	}
+miss:
+	return 0;
+}
+
+static inline int skb_ip_match(struct sk_buff *skb, struct filter *filter)
+{
+	struct iphdr iph;
+	union l4hdr l4h;
+
+	if (skb_l3hdr(skb, &iph, sizeof(struct iphdr))) {
+		goto miss;
+	}
+
+	if (!ip_addr_match(iph.saddr, iph.daddr, filter)
+	    || !proto_match(iph.protocol, filter)) {
+		goto miss;
+	}
+
+	if (skb_l4hdr(skb, &l4h, iph.ihl * 4)) {
+		goto miss;
+	}
+
+	return skb_l4_match(iph.protocol, &l4h, filter);
+miss:
+	return 0;
+}
+
+static inline int sk_is_ipv6(struct sock *sk)
+{
+	return sk->sk_family == AF_INET6;
+}
+
+static inline int skb_is_ipv6(struct sk_buff *skb)
+{
+	return skb->protocol && skb->protocol == htons(ETH_P_IPV6);
+}
+
+static inline int sk_is_ipv4(struct sock *sk)
+{
+	return sk->sk_family == AF_INET;
+}
+
+static inline int skb_is_ipv4(struct sk_buff *skb)
+{
+	return skb->protocol && skb->protocol == htons(ETH_P_IP);
+}
+
+static inline void filter_init(struct filter *filter)
+{
+	SADDR_FILTER
+	DADDR_FILTER
+	SPORT_FILTER
+	DPORT_FILTER 
+	PROTO_FILTER 
+	FLAGS_FILTER 
+	IP_VERSION_FILTER
+}
+
+static inline int sk_match_and_trace(struct pt_regs *ctx, struct sock *sk,
+				     union filter_args *args,
+				     int (*cb) (struct pt_regs *,
+						union filter_args *))
+{
+	struct filter filter = { };
+
+	filter_init(&filter);
+
+	if ((filter.ip_version == 0 || filter.ip_version == 4)
+	    && sk_is_ipv4(sk)) {
+		if (sk_ip_match(sk, &filter, args)) {
+			return cb(ctx, args);
+		}
+	} else if ((filter.ip_version == 0 || filter.ip_version == 6)
+		   && sk_is_ipv6(sk)) {
+		if (sk_ip6_match(sk, &filter, args)) {
+			return cb(ctx, args);
+		}
+	}
+
+	return 0;
+}
+
+static inline int skb_match_and_trace(struct pt_regs *ctx, struct sk_buff *skb,
+				      union filter_args *args,
+				      int (*cb) (struct pt_regs *,
+						 union filter_args *))
+{
+	struct filter filter = { };
+
+	filter_init(&filter);
+
+	if ((filter.ip_version == 0 || filter.ip_version == 4)
+	    && skb_is_ipv4(skb)) {
+		if (skb_ip_match(skb, &filter)) {
+			return cb(ctx, args);
+		}
+	} else if ((filter.ip_version == 0 || filter.ip_version == 6)
+		   && skb_is_ipv6(skb)) {
+		if (skb_ip6_match(skb, &filter)) {
+			return cb(ctx, args);
+		}
+	}
+
+	return 0;
+}
+
+static inline int __trace_sk_input(struct pt_regs *ctx, struct sock *sk)
+{
+	union filter_args args = { };
+
+	args.input = true;
+
```

Comment:
We can save a line here. Otherwise, the code looks too sparse. The same for other functions below.

Reply:
Emm, you are right.